presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -46,7 +47,16 @@
     @Override
     public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
     {
-        WindowNode child = captures.get(CHILD);
+        // Pulling the descendant WindowNode above projects is done as a part of this rule, as opposed in a
+        // separate rule, because that pullup is not useful on its own, and could be undone by other rules.
+        // For example, a rule could insert a project-off node between adjacent WindowNodes that use different
+        // input symbols.
```

Comment:
One minor correction to @maciejgrzybek :  FilterNode actually can't be moved, as changing the bag of rows could change the result of the window functions.  But project-on-like nodes such as SemiJoin, MarkDistinct, RowNumber, and AssignUniqueId could be moved.

An additional enhancement that could be made later would be to support moving the upper WindowNode down, as opposed to the lower WindowNode up.  This would support somewhat contrived situations.

@kokosing I don't see a reason to limit the pattern to zero-or-one project nodes between window nodes.  The algorithm would be no simpler, and the rule would be less generally effective.  Please keep in mind that the rule API exists for us to write rules -- it serves the rules, not the other way around.